### PR TITLE
Implement changes necessary for drift_crdt implementation.

### PR DESCRIPTION
**Pull request by JanezStupar**
These changes are a result of my work on an implementation of sqlite CRDT implementation for Drift (see: https://github.com/JanezStupar/drift/tree/sqlite_crdt/drift_crdt).

My work relies on the work of @cachapa (`sql_crdt` and `sqlite_crdt`) libraries and Simon Binder's `drift_sqflite` library.

The changes are mostly related to expansion of the support for the sqflite API within `sqlite_crdt` of the scope that was necessary to get the `drift_crdt` working.

Only new features are the migration within `sqlite_crdt.dart` and a serialization stub based on `json_serializable`.

I felt like those were appropriate places to dump the new features.

Let me know if you are open to accepting this contribution and what changes you feel need to be made to make them acceptable.

### CODE REVIEW

**Comment by cachapa** [comment]
This PR cannot be accepted due to the upstream PR not having been merged: https://github.com/cachapa/sql_crdt/pull/3

In any case I wanted to comment on a couple of points:

Raw access:
I made an explicit decision to not expose the underlying raw database because performing any changes without properly applying touching the HLCs can/will break the CRDT - in the best case everything stops working, but you're more likely to get subtle behavioural changes that result in incorrect merges and change propagation that are very hard to discover, much less debug.
You can obviously always just use standard SQL tooling to open the database directly and perform changes to your heart's content, but at least we're not implicitly encouraging it via our API.

Migrations:
One of the original versions actually had a migration mechanism as well, but I ended up removing it because I wanted it to be a very explicit operation.
Upgrading a normal database to a CRDT requires a lot of thought and care, not only on how the data will be handled locally, but it also requires a lot of thought on how it gets synced back to remote nodes.
I felt like that should require more thought than simply adding a `migrate: true` parameter.

Again, thank you for your contribution but I don't think these changes meet the project goals.

**Comment by JanezStupar** [comment]
First off I think you misunderstood the nature of the `rawUpdate`, `rawInsert`... methods. They are not meant to bypass the CRDT at all. They are merely a reflection of the sqflite APIs (see: https://pub.dev/documentation/sqflite/latest/sqlite_api/Database-class.html - e.g. `query` vs `rawQuery` methods)

In fact the API implemented in the `sqlite_crdt` is misnamed. The `query` method has the signature of the `rawQuery` method from Sqflite. The reason why I added them is because the client I am using is using the raw SQL statement API instead of parameterized `query` like API. 

However I agree that they should not be within `sql_crdt` library. But they cannot be implemented in `sqlite_crdt` - where they fit for sure, due to the private nature of the code within `base_crdt`.

Next issue is the isolation of the database connection to the `base_crdt`. While I can appreciate the sentiment I ran into an issue while implementing the transaction. Sqflite rolls back the transaction on error. `sql_crdt` implementation however does not and when trying to make `sqlite_crdt` compliant to the sqflite API this became an issue.

And since `sql_crdt` guards its implementation so closely I can not implement the required changes in superclass of TransactionCrdt. To do what is needed I would have to basically fork everything - to the extent of my knowledge.

And it seems to me these issues will crop up whenever somebody tries to include your libs within their existing project.

On the topic of migrations, I personally feel like that is a nice optional thing to have. Especially when it can be implemented in a database agnostic way. But that is a matter of personal preference. So by putting them inside the `sql_crdt` I though I was killing two birds with one stone. 

But the real issue is again, that since the database connection cannot be accessed by the downstream code it is not possible for my own database library (whatever that is) to extend the behaviour of `sqlite_crdt` in a pragmatic fashion. And I have noted your suggestion to open another database connection. However it is my understanding that in sqflite it is assumed there is only one open connection at any given time to the database at this point in time. I could be wrong on this one.

So in a nutshell it would seem to me that the decision to keep so much of the base stuff private is really detrimental to brownfield scenarios and I am pretty sure that other developers will be affected in the future. I can understand you find this approach elegant and it might make sense for your use case, but it has already costed me a lot of time to this point (not enough for me to warrant implementing CRDT from scratch, mind you) and while it has facilitated me learning about CRDTs by proxy (which is good too) it doesn't have to be like that.

Have you considered using the '@protected' decorator instead?

p.s.: I did some prototyping to see whether I missed something when originally implementing these changes. And I realized that without making CRDT code more accessible I really don't have an alternative.

**Comment by cachapa** [comment]
I see what you mean by the `rawQuery` methods. You're right that I misunderstood their intent.

> In fact the API implemented in the sqlite_crdt is misnamed.

It's not misnamed, it's only incidental that it has the same name as a method from the `sqflite` package. As I said before this library is intended to be a generic implementation of CRDTs backed by an SQL engine.
Moreover `sqlite_crdt` is an implementation of CRDTs backed by a Sqlite engine - not a `sqflite` engine - otherwise I would've named it `sqflite_crdt`.
The fact that `sqlite_crdt` uses `sqflite` is an implementation detail. It may change in the future (however unlikely).

> Next issue is the isolation of the database connection to the base_crdt. While I can appreciate the sentiment I ran into an issue while implementing the transaction. Sqflite rolls back the transaction on error. sql_crdt implementation however does not and when trying to make sqlite_crdt compliant to the sqflite API this became an issue.

Can you provide a bit more detail here? I'm under the impression that `sqlite_crdt` does roll back transactions. It uses the `sqflite` `transaction()` method which I believe handles cleanup.

> On the topic of migrations, I personally feel like that is a nice optional thing to have. Especially when it can be implemented in a database agnostic way. But that is a matter of personal preference. So by putting them inside the sql_crdt I though I was killing two birds with one stone.

That is a good point, and one I might come back to later even if at this moment I'm still trending to no.
The only suggestion I would make is to separate such pieces of work into separate PRs since it makes PRs smaller, easier to review and to focus the discussion.
I actually mention this in the [contribution guidelines](https://github.com/cachapa/sql_crdt/blob/master/CONTRIBUTING.md).

> But the real issue is again, that since the database connection cannot be accessed by the downstream code it is not possible for my own database library (whatever that is) to extend the behaviour of sqlite_crdt in a pragmatic fashion. And I have noted your suggestion to open another database connection. However it is my understanding that in sqflite it is assumed there is only one open connection at any given time to the database at this point in time. I could be wrong on this one.

Actually that was not a suggestion, just a statement of fact. I really don't think you should be accessing your CRDT databases using other tools unless it's read-only.
The correct way to access the underlying database is to update the version and use the `onUpgrade` callback to get a database object that's closer to bare metal.
Naturally this also opens the door to making invalid changes to the CRDT, but it does it in a limited space.

Maybe it would be helpful if you could provide examples of things you'd like to do with direct access to the underlying database that you can't do now.

To be clear I have migrated hundreds of thousands of users from brownfield applications to `sqlite_crdt` and `postgres_crdt` both in mobile apps and backend, and managed to do it with the exitsting API.
So I'm wondering if I'm missing some use case, or if perhaps the documentation isn't clear enough.

**Comment by JanezStupar** [comment]
>
> t's not misnamed, it's only incidental that it has the same name as a
> method from the sqflite package. As I said before this library is
> intended to be a generic implementation of CRDTs backed by an SQL engine.
>
> Yes, I have assumed too much without really giving it a thought based on
the `sqflite` dependency.

> Can you provide a bit more detail here? I'm under the impression that
> sqlite_crdt does roll back transactions. It uses the sqflite transaction()
> method which I believe handles cleanup.
>
 I went and revisited my transactions and tests. Yes you are absolutely
right. Apparently I misunderstood a test case in drift, which was related
to the drift architecture. And the explicit rollback helped pass the test
but did not address the actual issue.

> That is a good point, and one I might come back to later even if at this
> moment I'm still trending to no.
>
 Sure I can submit migrations as a separate PR, no problem. And might as
well keep them as part of the approach I am going to take going forward.

> The correct way to access the underlying database is to update the version
> and use the onUpgrade callback to get a database object that's closer to
> bare metal.
>
Duly noted.

> Maybe it would be helpful if you could provide examples of things you'd
> like to do with direct access to the underlying database that you can't do
> now.
>
 The ability to override transactions, like the use case above. Another one
would be to implement additional features like migrations (granted that
could be done using onUpgrade too). Those were two actual issues I ran
into. Also the missing database close method. But then again I can just
cache the database connection in my own APIs.

> So I'm wondering if I'm missing some use case, or if perhaps the
> documentation isn't clear enough.
>
Like I said above. Turns out I wanted to implement "`SqfliteCrdt`" based on
`SqliteCrdt` - which would be a reasonable thing to do, considering the
design scope (which I now understand better, thank you) of `sqlite_crdt`
and impedance mismatch towards my use case.

I understand now that the motive is to protect the CRDT integrity, which I
applaud. But the second order effect is that it makes a well meaning user -
that indeed doesn't want to fiddle with the CRDT directly, unnecessarily
complicated.

Going forward I decided to just fork the `sqlite_crdt` into `sqflite_crdt`.
But I would still like you to maybe accept a refactor of the `sql_crdt`
where I extracted some of the `timestamped_crdt.dart` into separate
functions - so I can use shared implementation in my own sqflite related
API (see the changes here:
https://github.com/JanezStupar/sqlite_crdt/blob/sqflite_crdt/lib/src/crdt_util.dart
).

In case you are open to that it is up to you whether you want these as a
part of `timestamped_crdt` or if you are fine with the static method
approach, or something else for that matter.

I have also found a couple bugs/edge cases that I will prepare a separate
PR's for.

Thank you for your time so far. And sorry for not reading the contributor
guidelines before going through this.

In my defence, this project I am on - using CRDT's to implement a shared
file system data sync solution for my app, went from "I wonder if this
would work, let me try it out" to "yeah this is the way to go" to finally
"let me see if I can share some of my changes" in a blur.

Thanks, looking forward to hearing from you.
-- 
Janez Štupar,
www.janezstupar.com

**Comment by cachapa** [comment]
> Sure I can submit migrations as a separate PR, no problem. And might as well keep them as part of the approach I am going to take going forward.

As I said, I'm not convinced of this approach yet. In fact I had this in a very early version of this package and removed it.
I wouldn't want you to invest more effort just to have a hanging PR while I make up my mind :-)

> The ability to override transactions, like the use case above.

I don't think there are any other sqflite mechanisms for transactions than the one we're using.
You're always able to issue `BEGIN TRANSACTION; ROLLBACK; COMMIT;` as raw SQL statements if you want, but honestly when I tried that I ran into a number of concurrency issues and came to the same conclusion that the sqflite dev came when he made his `transaction()` API (transactions should be exclusive and queued).

> Going forward I decided to just fork the `sqlite_crdt` into `sqflite_crdt`.

Would it be too much to ask that you use a more distinguishable name? I think it's not good for either project to have a single-letter difference in the title.

> But I would still like you to maybe accept a refactor of the `sql_crdt` where I extracted some of the `timestamped_crdt.dart` into separate functions - so I can use shared implementation in my own sqflite related API (see the changes here: https://github.com/JanezStupar/sqlite_crdt/blob/sqflite_crdt/lib/src/crdt_util.dart.

My understanding is that you'd like the opportunity to intercept statements after they've been "CRDT-fied". It's unclear for me why you would want to do that. My concerns remain: it makes implementations more complex and opens the door for subtly breaking the CRDT integrity.

It's actually against "well-meaning users" that I'm protecting against. Over the years of working on this I have been able to witness a number of edge cases where a very small error manifests in almost imperceptible discrepancies.
Over time I have implemented those learnings into this library, which is why it guards its internals so tightly.

In other words, I was the well-meaning user and I managed to break things despite my experience with building CRDTs and having written every single line of code for this package.

> I have also found a couple bugs/edge cases that I will prepare a separate PR's for.

Thanks!

> Thank you for your time so far. And sorry for not reading the contributor guidelines before going through this.

No worries. Thanks for taking the time for this discussion.

**Comment by JanezStupar** [comment]
>
> As I said, I'm not convinced of this approach yet. In fact I had this in a
> very early version of this package and removed it.
>
That is fine. I already resolved to add it to my fork.

> Going forward I decided to just fork the sqlite_crdt into sqflite_crdt.
>
> Would it be too much to ask that you use a more distinguishable name? I
> think it's not good for either project to have a single-letter difference
> in the title.
>
I do not object at all. FWIW, I blame it on the sqflite dev. If you have
any suggestions I am quite open to it.

> My understanding is that you'd like the opportunity to intercept
> statements after they've been "CRDT-fied". It's unclear for me why you
> would want to do that. My concerns remain: it makes implementations more
> complex and opens the door for subtly breaking the CRDT integrity.
>
No, I want nothing of the sort per say. What I want to do is essentially
what you have done - but conforming to the Sqflite API so I can plug it
into Drift.

With minimal discrepancies with the underlying implementation. Which would
be much easier if sql_crdt code was structured in a more modular fashion
(which by extension would make it more testable too).

What the refactoring I have proposed does is minimize the area that I have
to cover to keep in lockstep with the sql_crdt implementation as you make
changes going forward.

I could propose a higher degree of modularity and maybe even implement it.
But it seems that you prefer tighter coupling on your end.

And unless you are willing to give me any ground, then I will just have to
live with the approach I have taken currently, where I have to reimplement
portions of sql_crdt to accomplish my goals until one of us backs off.

For all I know I might realize in near future that my approach is not okay
and will resort to either dropping Drift or even forking it like you
suggested. But at this time I am comfortable with capabilities that I have
at my disposal on all ends. I would have preferred a more streamlined
solution, but that is neither here nor there and might change with time.

Thank you for your patience.

-- 
Janez Štupar,
www.janezstupar.com

**Comment by cachapa** [comment]
> I do not object at all. FWIW, I blame it on the sqflite dev. If you have any suggestions I am quite open to it.

Agreed, I understand his thought process but it's already let to some confusion on my side.
No suggestions though, sorry. Naming things is one of the hard problems in CS.

> No, I want nothing of the sort per say. What I want to do is essentially what you have done - but conforming to the Sqflite API so I can plug it into Drift.

Understood.
So for now I'm really weary of changing the architecture since it has the level of modularity I wanted.
The current implementation is working well and even manages to support two major DB backends without a lot of implementation effort from the integration side.
My goal is to implement more backends (e.g. MariaDB) so I'm trying to limit the effort and the surface for bugs when that time comes.

For now I will refrain from such changes. I'm looking forward to your feedback after you've had some experience running your fork.
